src/main/java/org/apache/sling/auth/oauth_client/impl/RedirectHelper.java

```diff
@@ -16,57 +16,135 @@
+        AuthenticationRequest.Builder authRequestBuilder = new AuthenticationRequest.Builder(
                 ResponseType.CODE,
-                clientID)
-                .scope(new Scope(scopes.toArray(new String[0])))
-                .endpointURI(authorizationEndpointUri)
-                .redirectionURI(redirectUri)
-                .state(state);
-
-        if (additionalAuthorizationParameters != null) {
-            additionalAuthorizationParameters.stream()
-                    .map( s -> s.split("=") )
-                    .filter( p -> p.length == 2 )
-                    .forEach( p -> authRequestBuilder.customParameter(p[0], p[1]));
+                scopes,
+                new ClientID(conn.clientId()),
+                redirectUri
+        )
+        .endpointURI(authorizationEndpointUri)
+        .state(state);
+
+        if (nonce != null) {
+            Cookie nonceCookie = buildCookie(path, OAuthStateManager.COOKIE_NAME_NONCE, nonce);
+            cookies.add(nonceCookie);
+
+            authRequestBuilder.nonce(new Nonce(nonce));
+
+        }
+
+        if (pkceEnabled) {
+            // Generate a new random 256 bit code verifier for PKCE
+            CodeVerifier codeVerifier = new CodeVerifier();
+
+            authRequestBuilder.codeChallenge(codeVerifier, CodeChallengeMethod.S256);
+
+            Cookie codeVerifierCookie = buildCookie(path, OAuthStateManager.COOKIE_NAME_CODE_VERIFIER, codeVerifier.getValue());
+            cookies.add(codeVerifierCookie);
+        }
+
+        if (originalRedirectTarget != null) {
+            Cookie redirectCookie = buildCookie(path, OAuthStateManager.COOKIE_NAME_REDIRECT_URI, originalRedirectTarget);
+            cookies.add(redirectCookie);
+        }
+
+        if (conn.additionalAuthorizationParameters() != null) {
+            conn.additionalAuthorizationParameters().stream()
+                    .map(s -> s.split("="))
+                    .filter(p -> p.length == 2)
+                    .forEach(p -> authRequestBuilder.customParameter(p[0], p[1]));
         }
         URI uri = authRequestBuilder.build().toURI();
-        return new RedirectTarget(uri, buildCookie(perRequestKey));
+        return new RedirectTarget(uri, cookies);
     }
+
     
-    private static @NotNull Cookie buildCookie(@NotNull String perRequestKey) {
-        Cookie cookie = new Cookie(OAuthStateManager.COOKIE_NAME_REQUEST_KEY, perRequestKey);
+    private static @NotNull Cookie buildCookie(@Nullable String path, @NotNull String name, @NotNull String perRequestKey) {
+        Cookie cookie = new Cookie(name, perRequestKey);
         cookie.setHttpOnly(true);
         cookie.setSecure(true);
         cookie.setMaxAge(COOKIE_MAX_AGE_SECONDS);
+        if (path !=null)
+            cookie.setPath(path);
         return cookie;
     }
+
+    public static String findLongestPathMatching(@NotNull String[] path,@Nullable String url) {
```

Comment:
at a first glance i didn't see this utility used outside of RedirectHelper -> make it private.

the return value can be null, right? -> annotate accordingly.

Reply:
i was mistaken -> resolving